pinot-core/src/main/java/org/apache/pinot/core/io/writer/impl/FixedByteSingleValueMultiColWriter.java

```diff
@@ -99,9 +100,12 @@ public void setBytes(int row, int col, byte[] bytes) {
     indexDataBuffer.readFrom(offset, bytes);
   }
 
+  @Override
   public void close()
       throws IOException {
-    this.indexDataBuffer.close();
-    this.indexDataBuffer = null;
+    if (this.indexDataBuffer != null) {
```

Comment:
You can make this final and remove the null check as PinotDataBuffer.close() is idempotent, your choice

Reply:
I had initially done that -- removed the setting to null and made it final. However, @mcvsubbu brought up a scenario in context of multi-threading to explain why we were setting to null intially. Looks like if two threads race in a particular way -- where one does a close and does not set it to null and soon after some other thread invokes a set***() operation on this class, then it will use the non-null buffer and invoke some set operation on it which may lead to arbitrary errors/bugs as opposed to hitting NPE if the first thread sets it to null. Based on this I reverted back to old behavior and made it non-final . 

I agree that PinotDataBuffer.close() is final and so we don't need the if check.

Reply:
In case of multi-threading, final is safer than non-volatile variable. I don't think we should throw NPE if the buffer is already closed, which could be confusing, and not align with Closable's usual behavior. I prefer the final + no null check way, @mcvsubbu what do you think?

Reply:
It is a balance here.

Pinot query processing threads are not interruptable (at least now). The parent thread lets the child thread run with the query while returning to the broker after some timeout. It is therefore possible that a query thread is using a segment (even after it has been released).

A retention or segment drop can cause the segment parts to be closed, and a read request may come along, causing us to access bad memory, and a JVM crash could result (we use direct access to memory, in unsafe manner)

NPE on a close is a much smaller evil, and I can live with that. 

In the early days we even saw JVM crashes during segment unload.

Reply:
One thought was that in case this multi-thread scenario does happen then NPE might be better than a core dump / segmentation fault (as a result of doing set operation on buffer that has been closed/released especially since all set/get operations on direct buffer are unsafe).

Reply:
This should not happen because we keep reference count on the segments. While it is true that if our code has bug then this could potentially cause JVM crash, so it is clean + fail fast model vs. try to survive failure model. I'm okay both way, but in FixedByteValueReaderWriter we are taking the first way, so I prefer keeping them consistent.

Reply:
@Jackie-Jiang, this can happen because refcnt is kept only until the end of the main thread of the query. Like I said, a child-thread could be running after the main thread reduces the refcnt. If there are other classes that are not following this, I suggest we bring them back to follow this model.

Reply:
Good point. Then let's make the data buffer volatile and set it to null when close it. Can you also change the fixed one to follow this model, and add some comments describing why we are doing this?

Reply:
I have made the changes in FixedByteSingleValueMultiColWriter, FixedByteSingleValueMultiColReader,FixedByteValueReaderWriter and FixedBitIntReaderWriter based on the model we agreed to.